runtime/precompiles/src/lib.rs

```diff
@@ -16,17 +16,170 @@
+			let deadline = offchain::timestamp();
+			deadline.add(Duration::from_millis(msec_cost));
+			offchain::sleep_until(deadline);
+		}
+
+		// TODO Should this actually be stopped?
+		// https://ethervm.io/#F3
+		// REvisit: does solidity void contract issue stop or return
+		Ok((ExitSucceed::Returned, [0u8; 0].to_vec(), gas_cost))
+	}
+}
+
+/// A precompile to wrap the functionality from parachain_staking::nominate.
+///
+/// EXAMPLE USECASE:
+/// A simple example usecase is a contract that allows donors to donate, and stakes all the funds
+/// toward one fixed address chosen by the deployer.
+/// Such a contract could be deployed by a collator candidate, and the deploy address distributed to
+/// supporters who want to donate toward a perpetual nomination fund.
+pub struct NominateWrapper<Runtime>(PhantomData<Runtime>);
+
+type BalanceOf<Runtime> = <<Runtime as parachain_staking::Config>::Currency as Currency<
+	<Runtime as frame_system::Config>::AccountId,
+>>::Balance;
+
+impl<Runtime> Precompile for NominateWrapper<Runtime>
+where
+	Runtime: parachain_staking::Config + pallet_evm::Config,
+	Runtime::AccountId: From<H160>,
+	BalanceOf<Runtime>: TryFrom<U256> + Debug,
+	Runtime::Call: Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+	<Runtime::Call as Dispatchable>::Origin: From<Option<Runtime::AccountId>>,
+	Runtime::Call: From<parachain_staking::Call<Runtime>>,
+{
+	fn execute(
+		input: &[u8], //Reminder this is big-endian
+		target_gas: Option<u64>,
+		context: &Context,
+	) -> Result<(ExitSucceed, Vec<u8>, u64), ExitError> {
+		// Basic sanity checking for length
+		// https://solidity-by-example.org/primitives/
+		const COLLATOR_SIZE_BYTES: usize = 20;
+		const AMOUNT_SIZE_BYTES: usize = 32;
+		const TOTAL_SIZE_BYTES: usize = COLLATOR_SIZE_BYTES + AMOUNT_SIZE_BYTES;
+
+		if input.len() != COLLATOR_SIZE_BYTES + AMOUNT_SIZE_BYTES {
+			return Err(ExitError::Other(
+				"input length for Sacrifice must be exactly 16 bytes".into(),
+			));
+		}
+
+		// Convert to right data types
+		//TODO I guess we should use the account mapping if possible. Otherwise this won't work for
+		// chains tht have a more standard substrate account type.
+		let collator = H160::from_slice(&input[0..COLLATOR_SIZE_BYTES]);
+
+		let amount: BalanceOf<Runtime> =
+			sp_core::U256::from_big_endian(&input[COLLATOR_SIZE_BYTES..TOTAL_SIZE_BYTES])
+				.try_into()
+				.map_err(|_| {
+					ExitError::Other("amount is too large for Runtime's balance type".into())
+				})?;
+
+		println!("Collator account is {:?}", collator);
+		println!("Amount is {:?}", amount);
+
+		// Construct a call
+		let inner_call = parachain_staking::Call::<Runtime>::nominate(collator.into(), amount);
+		let outer_call: Runtime::Call = inner_call.into();
+		let info = outer_call.get_dispatch_info();
+
+		// Make sure enough gas
+		if let Some(gas_limit) = target_gas {
+			let valid_weight = info.weight <= Runtime::GasWeightMapping::gas_to_weight(gas_limit);
```

Comment:
This comparison, which is also present in Frontier, assumes a monotonic weight gas mapping. While that is probably the most likely, it should be noted in the docs for the weight gas mapping.

It would probably be more correct, and remove the monotonic requirement to map the call's weight to gas and then compare gases.

Reply:
Fixed in cedc825c1e9c7736d91a6d51bea51225f74fb3b4 and reported to frontier in https://github.com/paritytech/frontier/issues/369